### PR TITLE
Source Github: implement client-side throttling of requests

### LINKED ISSUE #5036: Source Github: implement client-side throttling of requests
## Tell us about the problem you're trying to solve
related to #2742 

Sometimes a user wants to use up only N% of their API quota. For example, GH allows only 5000 requests per hour, and the token may be shared between Jenkins and Airbyte. in that case the user doesn't want Airbyte to use up more than say 30% of the quota so it doesn't block jenkins. 

## Describe the solution you’d like
A parameter in the connector configuration which controls how much of the API quota it uses. 

**Urgency:** Nice-to-have

### PR DESCRIPTION
**Pull request by grubberr**
## What
Issue: https://github.com/airbytehq/airbyte/issues/5036

Implement new `MultipleTokenAuthenticatorWithRateLimiter` if customer provided `requests_per_hour` param
- For authentication use a similar implementation for `MultipleTokenAuthenticator`.
- RATE-LIMITING: Every token in the cycle is checked against the rate limiter and if the token is over capacity switch to another token. If all tokens are exhausted sleep until the first token will be ready.

## Pre-merge Actions
*Expand the relevant checklist and delete the others.*

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-github/source_github/source.py [hunk 1]
```diff
@@ -152,6 +153,13 @@ def _get_authenticator(config: Dict[str, Any]):
             creds = config.get("credentials")
             token = creds.get("access_token") or creds.get("personal_access_token")
         tokens = [t.strip() for t in token.split(TOKEN_SEPARATOR)]
+        requests_per_hour = config.get("requests_per_hour")
+        if requests_per_hour:
+            return MultipleTokenAuthenticatorWithRateLimiter(
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-github/source_github/source.py:158]: Let's add unit tests for `MultipleTokenAuthenticatorWithRateLimiter` usage.

**Review by lazebnyi — APPROVED** [review]
LGTM! Need add small unit tests.

File: airbyte-integrations/connectors/source-github/source_github/source.py (hunk 1 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-github/source_github/source.py:158]: I have added unit_test for MultipleTokenAuthenticatorWithRateLimiter class

**Comment by grubberr** [comment]
@lazebnyi do we pass this to Airbyte review or we can finish it ourselves?

**Comment by lazebnyi** [comment]
> @lazebnyi do we pass this to Airbyte review or we can finish it ourselves?

We must send this PR to Airbyte review due to spec changes.

File: airbyte-integrations/connectors/source-github/source_github/utils.py [hunk 2]
```diff
@@ -24,3 +30,57 @@ def read_full_refresh(stream_instance: Stream):
...
+    @property
+    def auth_header(self) -> str:
+        return self._auth_header
+
+    @property
+    def token(self) -> str:
+        while True:
+            token = next(self._tokens_iter)
+            if self._check_token(token):
+                return f"{self._auth_method} {token}"
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-github/source_github/utils.py:62]: I'm not sure to understand how you perform:
> If a token exceeds the capacity limit, the system switches to another token.

From reading the code I've the impression that the execution is blocked sleeping when a token is exhausted and that you're not returning the next one that might not be exhausted.
I might be wrong. But I think your code would deserver additional docstring and testing 😄 .

File: airbyte-integrations/connectors/source-github/unit_tests/test_source.py [hunk 3]
```diff
@@ -165,3 +169,24 @@ def test_config_validation(repos_config, expected):
 def tests_get_and_prepare_repositories_config(config, expected):
     actual = SourceGithub._get_and_prepare_repositories_config(config)
     assert actual == expected
+
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-github/unit_tests/test_source.py:172]: I'd appreciate a test to check that the token rotation logic works correctly.

**Review by alafanechere — CHANGES_REQUESTED** [review]
Thank you for this interesting feature addition! Please  make a minor version increment. I suggest to add more docstrings and tests to make sure the logic implementation is correct and future maintainer can understand it.  Moreover, for additional Connector Acceptance Test coverage I sugggest to add a new config with multiple tokens to GSM and add tests with it in `acceptance-test-config.yaml`

**Comment by alafanechere** [comment]
@roman-yermilov-gl here's the [breaking change release plan draft](https://docs.google.com/document/d/1zWtBOmn85IqxecgXpNHY_72T3ZubsGGGUN8skxByH2c/edit#). I'd need more context on why this change is introduced to move forward.

File: airbyte-integrations/connectors/source-github/unit_tests/test_source.py (hunk 3 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-github/unit_tests/test_source.py:172]: I have improved unit test

File: airbyte-integrations/connectors/source-github/source_github/utils.py (hunk 2 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-github/source_github/utils.py:62]: I have improved unit test, now test shows that we take another token if first one is exhausted
I have added docs for `def _sleep`

**Comment by grubberr** [comment]
@alafanechere GSM already contains multiple token config

File: airbyte-integrations/connectors/source-github/Dockerfile [hunk 4]
```diff
@@ -12,5 +12,5 @@ RUN pip install .
 ENV AIRBYTE_ENTRYPOINT "python /airbyte/integration_code/main.py"
 ENTRYPOINT ["python", "/airbyte/integration_code/main.py"]
 
-LABEL io.airbyte.version=0.4.10
+LABEL io.airbyte.version=0.4.11
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-github/Dockerfile:15]:
```suggestion
LABEL io.airbyte.version=0.5.0
```

You're adding a new feature, so a minor bump is better 😄

File: airbyte-integrations/connectors/source-github/metadata.yaml [hunk 5]
```diff
@@ -5,7 +5,7 @@ data:
   connectorSubtype: api
   connectorType: source
   definitionId: ef69ef6e-aa7f-4af1-a01d-ef775033524e
-  dockerImageTag: 0.4.10
+  dockerImageTag: 0.4.11
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-github/metadata.yaml:8]:
```suggestion
  dockerImageTag: 0.5.0
```

File: airbyte-integrations/connectors/source-github/unit_tests/test_source.py [hunk 6]
```diff
@@ -2,13 +2,20 @@
...
 from unittest.mock import MagicMock
 
 import pytest
 import responses
 from airbyte_cdk.models import AirbyteConnectionStatus, Status
+<<<<<<< HEAD
+from freezegun import freeze_time
+=======
 from airbyte_cdk.utils.traced_exception import AirbyteTracedException
+>>>>>>> master
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-github/unit_tests/test_source.py:16]: I think the conflict fix went no well 😄

File: docs/integrations/sources/github.md [hunk 7]
```diff
@@ -163,15 +164,16 @@ The GitHub connector should not run into GitHub API limitations under normal usa
 
 | Version | Date       | Pull Request                                                                                                      | Subject                                                                                                                                                             |
 |:--------|:-----------|:------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| 0.4.11  | 2023-05-16 | [25793](https://github.com/airbytehq/airbyte/pull/25793)                                                          | Implement client-side throttling of requests                                                                                                                        |
```
Comment (alafanechere) [inline docs/integrations/sources/github.md:167]:
```suggestion
| 0.5.0  | 2023-05-16 | [25793](https://github.com/airbytehq/airbyte/pull/25793)                                                          | Implement client-side throttling of requests                                                                                                                        |
```

**Review by alafanechere — CHANGES_REQUESTED** [review]
Thank you for the improved unit test.
I spotted a conflict resolution error. Please also re-run /test to make sure your new unit test is ran.

**Review by alafanechere — APPROVED** [review]
Nice!

**Pull request merged**